### ISSUE TITLE
Normalization issue with `#` (`%23`)

### ISSUE THREAD

**Issue by matthieuprat** [sporkmonger/addressable#295]
Not sure this is an actual bug:

```rb
Addressable::URI.parse("http://example.org?foo=%E9%23").normalize.query 
```

This returns the string `foo=%E9#`. I would have expected `foo=%E9%23`.

Note that `%E9` is the escaped version of the `é` character in ISO-8859-1, that is `URI.escape('é'.encode('iso-8859-1'))`.

Is this the intended behavior?

**Comment by dentarg** [comment]
No answers, but I found that `#query` and `#query_values` doesn't match:

```ruby
$ irb -raddressable/uri
irb(main):001:0> Addressable::VERSION::STRING
=> "2.5.2"
irb(main):002:0> Addressable::URI.parse("http://example.org?foo=%E9%23").normalize.query
=> "foo=%E9#"
irb(main):003:0> Addressable::URI.parse("http://example.org?foo=%E9%23").normalize.query_values
=> {"foo"=>"\xE9#"}
irb(main):004:0> Addressable::URI.unencode("%E9%23")
=> "\xE9#"
```

I think this issue is similar to https://github.com/sporkmonger/addressable/issues/224

**Comment by sporkmonger** [comment]
Yeah, if `query` and `query_values` don't match, that feels buggy to me.

**Comment by sporkmonger** [comment]
https://tools.ietf.org/html/rfc3986#section-2.5 applies here.

**Comment by sporkmonger** [comment]
We should probably add a test case for `uri.query_values = {"À": "ア"}` since its a cited example.

**Comment by dentarg** [comment]
Related to #334

**Comment by dentarg** [comment]
> I found that `#query` and `#query_values` doesn't match

Probably due to https://github.com/sporkmonger/addressable/issues/114#issuecomment-16338080

> Ultimately, the `query_values` method is attempting to emulate the `application/x-www-form-urlencoded` content type, poorly specified though it may be.

**Comment by dentarg** [comment]
> This returns the string `foo=%E9#`. I would have expected `foo=%E9%23`.

I think this is another variant of https://github.com/sporkmonger/addressable/issues/366 where addressable incorrectly decodes the percent-encoded reserved character `#` (`%23`)
